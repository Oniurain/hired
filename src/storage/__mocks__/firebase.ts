import * as fetchMock from 'fetch-mock'

fetchMock.get('http://localhost:8000/auth/me', {token: '1234-5678'})

let firebaseSignInWithCustomTokenMock = jest.fn(() => true)

const firebaseSetMock = jest.fn()

const initializeApp = () => {
  return {
    database() {
      return {
        once: jest.fn(),
        ref() {
          return {
            set: firebaseSetMock
          }
        }
      }
    }
  }
}

const auth = () => {
  return {
    signInWithCustomToken: firebaseSignInWithCustomTokenMock
  }
}

const setCustomTokenMock = (returnValue: boolean) => {
  firebaseSignInWithCustomTokenMock = jest.fn(() => returnValue)
}

export {
  auth,
  firebaseSetMock,
  firebaseSignInWithCustomTokenMock,
  initializeApp,
  setCustomTokenMock
}