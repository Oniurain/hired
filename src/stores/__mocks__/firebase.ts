import * as fetchMock from 'fetch-mock'

fetchMock.get('http://localhost:8000/auth/me', {token: '1234-5678'})

export const firebaseSetMock = jest.fn()

export const firebaseSignInWithCustomTokenMock = jest.fn(() => Promise.resolve(true))

export function initializeApp() {
  return {
    database() {
      return {
        ref() {
          return {
            set: firebaseSetMock
          }
        }
      }
    }
  }
}

export function auth() {
  return {
    signInWithCustomToken: firebaseSignInWithCustomTokenMock
  }
}
