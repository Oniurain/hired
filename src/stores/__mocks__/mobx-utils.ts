const mobxUtils = jest.genMockFromModule('mobx-utils')

mobxUtils.now = () => (+new Date(2017, 1, 14, 15, 22))

module.exports = mobxUtils