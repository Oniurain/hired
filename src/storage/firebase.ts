import * as firebase from 'firebase';

const config = {
  apiKey: 'AIzaSyDt83bvU6CXetNYGRWuaI12nLxpEbytR6I',
  authDomain: 'mobx-time-tracking.firebaseapp.com',
  databaseURL: 'https://mobx-time-tracking.firebaseio.com',
  storageBucket: 'mobx-time-tracking.appspot.com',
  messagingSenderId: '866097267685'
};
firebase.initializeApp(config);

const database = firebase.database();
const entries = firebase.database().ref('entries');

const Fb = {
  database,
  entries
};

export { Fb };

export interface IStorageObject {
  startTime: number;
  endTime: number|null;
  running: Boolean;
}
