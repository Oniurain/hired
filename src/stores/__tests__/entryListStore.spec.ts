import * as mobx from 'mobx'

import { firebaseSetMock } from '../../storage/__mocks__/firebase'

import FirebaseRepository from '../../storage/firebaseRepository'
import { RootStore, EntryListStore } from '../'

import { MOCK_ENTRIES } from '../../test/mockData'

describe('EntryListStore', () => {
  let rootStoreMock: RootStore
  let subject: EntryListStore

  beforeEach(() => {
    mobx.extras.resetGlobalState()
    jest.mock('firebase')
    firebaseSetMock.mockReset()

    const repository = new FirebaseRepository()
    rootStoreMock = new RootStore(repository, 'me')

    subject = new EntryListStore(
      rootStoreMock
    )
  })

  it('should initialize', () => {
    expect(subject).toBeDefined
    expect(mobx.toJS(subject.entries)).toEqual([])
    expect(subject.active).toBe(null)
  })

  describe('Hydration', () => {
    it('should hydrate from stored data', () => {
      subject.hydrate(MOCK_ENTRIES)

      expect(mobx.toJS(subject.entries)).toMatchSnapshot()
      expect(subject.active).toBe(null)
    })

    it('should sort all entries newest first', () => {
      subject.hydrate(MOCK_ENTRIES)

      expect(subject.entries[0].id).toEqual(MOCK_ENTRIES[3].id)
      expect(subject.entries[1].id).toEqual(MOCK_ENTRIES[2].id)
      expect(subject.entries[2].id).toEqual(MOCK_ENTRIES[1].id)
      expect(subject.active).toBe(null)
    })

    it('should set the newest entry active if it does not have an endTime', () => {
      const LOCAL_TEST_DATA = [{
        id: '5555-6666-7777',
        startTime: +new Date(2017, 1, 28, 9, 0),
        endTime: null,
        rate: 35.50,
        projectId: '1234-5678-9012'
      }, ...MOCK_ENTRIES]

      subject.hydrate(LOCAL_TEST_DATA)

      expect(subject.active).not.toBe(null)
      expect(subject.active.id).toBe('5555-6666-7777')
    })
  })
})
