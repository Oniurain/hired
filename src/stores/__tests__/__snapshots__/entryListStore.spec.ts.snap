// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryListStore Hydration should hydrate from stored data 1`] = `
Array [
  Object {
    "endTime": 2017-02-17T10:22:00.000Z,
    "id": "5346-1234-1243",
    "projectId": "1234-5678-9012",
    "rate": 35.5,
    "startTime": 2017-02-17T02:45:00.000Z,
  },
  Object {
    "endTime": 2017-02-16T19:59:00.000Z,
    "id": "3456-3456-3244",
    "projectId": "1234-5678-9012",
    "rate": 35.5,
    "startTime": 2017-02-16T10:22:00.000Z,
  },
  Object {
    "endTime": 2017-02-15T17:21:00.000Z,
    "id": "9786-2345-1234",
    "projectId": "1234-5678-9012",
    "rate": 35.5,
    "startTime": 2017-02-15T09:59:00.000Z,
  },
  Object {
    "endTime": 2017-02-14T14:59:00.000Z,
    "id": "9845-6524-2435",
    "projectId": "2345-9384-324523",
    "rate": 100.5,
    "startTime": 2017-02-14T07:23:00.000Z,
  },
  Object {
    "endTime": 2017-02-14T14:59:00.000Z,
    "id": "4567-2341-2369-7684",
    "projectId": "2345-9384-324523",
    "rate": 15.5,
    "startTime": 2017-02-14T07:23:00.000Z,
  },
  Object {
    "endTime": 2017-02-14T14:59:00.000Z,
    "id": "2342-6552-5345",
    "projectId": "1234-5678-9012",
    "rate": 35.5,
    "startTime": 2017-02-14T07:23:00.000Z,
  },
]
`;
